### PR TITLE
Adds NodeReady status condition to IngressController

### PR DESCRIPTION
**Pull request by danehans**
Many `ingresscontroller` failure scenarios can be traced back to a lack of available worker nodes. This PR introduces the `NodeReady` condition type, indicating whether the number of required nodes are available to support `ingresscontroller.spec.replicas`.

__Note:__ The type's name is not prefaced with `Worker` since the limitation for running ingress controllers on masters will be removed in the future.

/cc @Miciah @ironcladlou @knobunc

### CODE REVIEW

File: operator/v1/types_ingress.go [hunk 1]
```diff
@@ -362,6 +362,8 @@ var (
 	// DNSReady indicates the ready state of any DNS records for the ingress
 	// controller.
 	DNSReadyIngressConditionType = "DNSReady"
+	// NodeReady indicates the node state for the ingress controller.
+	NodeReadyIngressConditionType = "NodeReady"
```
Comment (Miciah) [inline operator/v1/types_ingress.go]: We've been going towards the pattern `FooDegraded`; how about `NodeDegraded`?

**Comment by ironcladlou** [comment]
In this context the operator really only knows about the deployment and its status (progressing, failed, conditions, etc.). Let's say like in this case the deployment is stalled because not enough pods are scheduled. How can our operator figure this out without encapsulating all the logic from layers beneath the deployment (like the scheduler)? Where would this end? For example, what if nodes are ready, but are tainted? Or mislabelled? Or any other number of states which could have driven the scheduling decisions. Would every operator need to do the same analysis? Many of them also manage deployments which have exactly the same failure modes.

Basically it seems like an abstraction leak and makes me wonder if there's some other approach to surface something meaningful...

**Comment by danehans** [comment]
@ironcladlou thanks for the feedback. The kube-scheduler operator currently surfaces status conditions for master nodes. If 2 worker nodes are required for a default cluster, maybe it makes sense for kube-scheduler operator to expand status conditions to cover worker nodes?

**Comment by danehans** [comment]
xref: https://github.com/openshift/cluster-kube-scheduler-operator/issues/201

File: operator/v1/types_ingress.go (hunk 1 above)
Comment (danehans) [reply operator/v1/types_ingress.go]: OK. I'll update after we settle on the approach for solving this problem.

**Comment by ironcladlou** [comment]
cc @deads2k

**Comment by rphillips** [comment]
I don't love the 'NodeDegraded term here when it is really the worker pool that might be degraded.

**Comment by smarterclayton** [comment]
/hold

This is definitely a key issue, we need a bit larger consensus (holding to get the discussion, not because it's a bad idea)

**Comment by smarterclayton** [comment]
This is probably more of a `Reason` than a Condition.

**Comment by smarterclayton** [comment]
Overall I would expect Degraded cluster operators that can't make progress to report Degraded Reason=NodeNotReady.  That might be something we can easily fix in Daemonset controller to show up as a reason on Available and Progressing.  If ingress controller could report the reason, that's a great start.  A separate condition may not make sense in all cases (if you know it's node not ready, isolating that helps).

**Comment by Miciah** [comment]
For the ingress operator, we have started to adopt a pattern whereby the ingresscontroller has fine-grained status conditions: `DeploymentDegraded`, `LoadBalancerReady`, and so on.  These conditions are aggregated to compute the ingresscontroller's overall `Available` and `Degraded` condition.  The ingresscontroller's `Available` and `Degraded` conditions are in turn aggregated to compute the clusteroperator's status conditions.  So ultimately the clusteroperator will report `Degraded` with a list of degraded ingresscontrollers, and the ingresscontroller will report `Degraded` with the specific reason (or perhaps a list of reasons), such as `DeploymentUnavailable` or perhaps `NodeNotReady`.

**Comment by ironcladlou** [comment]
>For the ingress operator, we have started to adopt a pattern whereby the ingresscontroller has fine-grained status conditions: DeploymentDegraded, LoadBalancerReady, and so on. These conditions are aggregated to compute the ingresscontroller's overall Available and Degraded condition. The ingresscontroller's Available and Degraded conditions are in turn aggregated to compute the clusteroperator's status conditions. So ultimately the clusteroperator will report Degraded with a list of degraded ingresscontrollers, and the ingresscontroller will report Degraded with the specific reason (or perhaps a list of reasons), **such as DeploymentUnavailable or perhaps NodeNotReady.**

DeploymentUnavailable makes sense because it's something the operator absolutely knows — the operator owns the deployment, the API on deployment gives us a somewhat opaque but semantically unambiguous "unavailable" condition.

It's the leap to "the deployment is unavailable because NodeNotReady" which surprises me, as it's the operator making an inference about the deployment's opaque condition that only lower layers actually know.

I'm not saying the operator _can't_ guess (and even guess correctly), but it smells because it _seems_ to be a layer violation that implies _all operators using the deployment resource_ would have to re-encode the same heuristics.

We already propagate the deployment's `Available` reason, which seems appropriate. If the reason is too opaque to be useful, it seems like the problem needs solved in the layers of ownership under the deployment.

**Comment by Miciah** [comment]
> DeploymentUnavailable makes sense because it's something the operator absolutely knows — the operator owns the deployment, the API on deployment gives us a somewhat opaque but semantically unambiguous "unavailable" condition.

We could scan events for `SchedulingFailed` and report that, possibly with guesses based on node statuses.

> It's the leap to "the deployment is unavailable because NodeNotReady" which surprises me, as it's the operator making an inference about the deployment's opaque condition that only lower layers actually know.

Ingress has some domain specificity, in the its operand uses affinity rules; most other operators probably do not set affinity rules on their operands.

> We already propagate the deployment's `Available` reason, which seems appropriate. If the reason is too opaque to be useful, it seems like the problem needs solved in the layers of ownership under the deployment.

That would be preferable indeed.

**Comment by ironcladlou** [comment]
@Miciah good points — no argument that we are well positioned to make good guesses, and that it might be useful to do so even if we shouldn't have to...

However, does that necessitate new condition API types and reasons, or would improved message strings for existing condition types be sufficient?

**Comment by Miciah** [comment]
One of our goals was to keep conditions simple: each condition should represent either a single observation or a synthesis of other conditions.  However, it is true that `DeploymentDegraded` will be set in this scenario, so maybe it would make sense to re-use that condition—the `DeploymentDegraded` condition's status would still reflect the deployment's `Available` condition, but when the deployment were not available, we could do some additional diagnostics to infer a probable reason from events and nodes.

**Comment by smarterclayton** [comment]
Given evidence today, a large majority of failed deployments / cluster operator unavailable are simply a down node (1.5% of clusters have a down node).  So if you can unambiguously or can reasonably pin that down, your Degraded=true reason should be NodeNotReady.

**Comment by deads2k** [comment]
> Given evidence today, a large majority of failed deployments / cluster operator unavailable are simply a down node (1.5% of clusters have a down node). So if you can unambiguously or can reasonably pin that down, your Degraded=true reason should be NodeNotReady. That we can then filter from our results and see a much more nuanced view.

This popped back up to front of mind because of metal problems that we hit several time where nodes didn't come up.  A reason is probably enough, but I'd really like see such a reason with a very nice message to make this more obvious to users.
